### ISSUE TITLE
Consider using memset to initialize the initial state of objects

### ISSUE THREAD

**Issue by DanHeidinga** [qbicc/qbicc#637]
Currently, qbicc initializes all fields in objects individually and sets the header field values in `initializeObjectHeader()`.  This does a field by field walk through the type hierarchy to set each field to 0.

Eventually, this will be done in conjunction wth a GC and likely use a "TLH/TLAB" (Thread local heap / thread local allocation buffer) that can be initialized to 0 in bulk.  For now, it likely be more efficient to set use memset to initialize the object fields and only explicitly write the non-0 values (like the typeid in the header).

Initialize header fields:
https://github.com/qbicc/qbicc/blob/591a429e7354a64cad08bf637acd306a0ec92837/plugins/gc/nogc/src/main/java/org/qbicc/plugin/gc/nogc/NoGcBasicBlockBuilder.java#L136-L143

Initialize instance fields:
https://github.com/qbicc/qbicc/blob/591a429e7354a64cad08bf637acd306a0ec92837/plugins/gc/nogc/src/main/java/org/qbicc/plugin/gc/nogc/NoGcBasicBlockBuilder.java#L55-L64

**Comment by dgrove-oss** [comment]
@dmlloyd had requested field-by-field zero initialization so that LLVM would understand what we were doing.

**Comment by DanHeidinga** [comment]
I'm not sure I follow.  Doesn't LLVM understand `memset`?  It must for compiling C/C++.  Is it related to our use of StatePoint?  I'll leave it at I'm confused and see what @dmlloyd can remember about the rationale for this

**Comment by dmlloyd** [comment]
Originally we had a `zeroinitializer` store for the whole allocated object, but LLVM generally can't lower that as a store, particularly if atomics are in play, so in the interests of simplicity, this was broken up. Within our IR, having a "store" means that our use-def chains would always trivially have an initial store to tie loads and subsequent stores to, making it particularly easy to eliminate redundant stores (perhaps even the entire initialization store).  But given that we do not yet have much in terms of an implementation of use-def chains, we could potentially model this another way, for example as a `zeroinitializer` store of the entire object as we had before.  Given that  `clang` [does in fact use memset](https://godbolt.org/z/rjK3E3aP8), at least for non-atomics, this could then be lowered to [the memset intrinsic](https://llvm.org/docs/LangRef.html#llvm-memset-intrinsics) (this involves a bitcast; see the compiler explorer link above).  It is my belief that the `memset` call would have to be followed by an `unordered` fence before it can be allowed to be observed (even within the same thread), which might be (but probably is not) more costly than having `unordered` on each zero write.

I think clearing in this way might make it slightly more difficult to detect cases where most or all fields are initialized before the original zero value can be observed, but it's probably OK for now regardless. If we get to a point where redundant initialization actually matters in practice, we could revisit.

TL;DR the original reasons were good but changing to `memset` intrinsics is probably fine also.
